core/include/detray/tools/rk_stepper.hpp

```diff
@@ -0,0 +1,172 @@
+     *
+     * @return returning the heartbeat, indicating if the stepping is alive
+     */
+    DETRAY_HOST_DEVICE
+    bool step(state& stepping, scalar es) {
+        auto& sd = stepping._step_data;
+
+        scalar error_estimate = 0;
+
+        // First Runge-Kutta point
+        sd.k1 = evaluatek(stepping, sd.b_first, 0);
+
+        const auto try_rk4 = [&](const scalar& h) -> bool {
+            // State the square and half of the step size
+            const scalar h2 = h * h;
+            const scalar half_h = h * 0.5;
+            // const auto pos = stepping().pos();
+            // const auto dir = stepping().dir();
+
+            // Second Runge-Kutta point
+            // const vector3 pos1 = pos + half_h * dir + h2 * 0.125 * sd.k1;
+            // sd.B_middle = getField(state.stepping.stepping, pos1);
+            sd.k2 = evaluatek(stepping, sd.b_middle, 1, half_h, sd.k1);
+
+            // Third Runge-Kutta point
+            sd.k3 = evaluatek(stepping, sd.b_middle, 2, half_h, sd.k2);
+
+            // Last Runge-Kutta point
+            // const vector3 pos2 = pos + h * dir + h2 * 0.5 * sd.k3;
+            // sd.B_last = getField(state.stepping.stepping, pos2);
+            sd.k4 = evaluatek(stepping, sd.b_last, 3, h, sd.k3);
+
+            // Compute and check the local integration error estimate
+            // @Todo
+            const auto err_vec = h2 * (sd.k1 - sd.k2 - sd.k3 + sd.k4);
+            error_estimate =
+                std::max(getter::norm(err_vec), static_cast<scalar>(1e-20));
+
+            return (error_estimate <= stepping._tolerance);
+        };
+
+        scalar step_size_scaling = 1.;
+        size_t n_step_trials = 0;
+
+        while (!try_rk4(stepping._step_size)) {
+            step_size_scaling = std::min(
+                std::max(0.25,
+                         std::sqrt(std::sqrt((stepping._tolerance /
+                                              std::abs(2. * error_estimate))))),
+                4.);
+
+            stepping._step_size = stepping._step_size * step_size_scaling;
+
+            // If step size becomes too small the particle remains at the
+            // initial place
+            if (stepping._step_size * stepping._step_size <
+                stepping._step_size_cutoff * stepping._step_size_cutoff) {
+                // Not moving due to too low momentum needs an aborter
+                printf("step size is too small. will break. \n");
+                return false;
+            }
+
+            // If the parameter is off track too much or given step_size is not
+            // appropriate
+            if (n_step_trials > stepping._max_rk_step_trials) {
+                // Too many trials, have to abort
+                printf("too many rk4 trials. will break. \n");
+                return false;
+            }
+            n_step_trials++;
+        }
+
+        stepping._step_size = std::min(stepping._step_size, es);
```

Comment:
@niermann999 Am I using `es` correctly here? I guess the step size should be smaller or equal to `es`?